docs/architecture/adr-071-crypto-v2-multi-curve.md

```diff
@@ -0,0 +1,804 @@
+    metadata := provider.Metadata()
+    id := generateProviderID(metadata)
+    providerFactories[id] = factory
+}
+
+// CreateCryptoProviderFromRecordOrConfig creates a CryptoProvider based on the provided Record or ProviderMetadata.
+// It enforces that exactly one of Record or ProviderMetadata must be provided.
+func CreateCryptoProviderFromRecordOrConfig(id string, record *Record, config *ProviderMetadata) (CryptoProvider, error) {
+    factory, exists := providerFactories[id]
+    if !exists {
+        return nil, fmt.Errorf("no factory registered for id %s", id)
+    }
+
+    // Validate input parameters
+    if record == nil && config == nil {
+        return nil, fmt.Errorf("both record and config cannot be nil")
+    }
+    if record != nil && config != nil {
+        return nil, fmt.Errorf("both record and config cannot be provided simultaneously")
+    }
+
+    // Determine which factory method to call based on the input
+    if record != nil {
+        return factory.CreateFromRecord(record)
+    }
+    if config != nil {
+        return factory.CreateFromConfig(*config)
+    }
+
+    // This line should never be reached due to the checks above
+    return nil, fmt.Errorf("unexpected error in CreateCryptoProviderFromRecordOrConfig")
+}
+
+// generateProviderID is a function that generates a unique identifier for a CryptoProvider based on its metadata.
+// This can be changed in the future to a more suitable function if needed.
+func generateProviderID(metadata ProviderMetadata) string {
+    return fmt.Sprintf("%s-%s", metadata.Type, metadata.Version)
+}
+```
+
+**Example**: Ledger HW implementation 
+
+Below is an example implementation of how a Ledger hardware wallet `CryptoProvider` might implement the registration of its factory and how instantiation would work.
+
+```go
+// crypto/v2/providers/ledger/factory.go
+
+type LedgerCryptoProviderFactory struct {
+    DevicePath string
+    // Any other necessary fields goes here 
+}
+
+func (f *LedgerCryptoProviderFactory) CreateFromRecord(record *Record) (CryptoProvider, error) {
+    // Extract necessary data from the record to initialize a LedgerCryptoProvider
+    if record == nil {
+        return nil, fmt.Errorf("record is nil")
+    }
+
+    // Assuming the record contains necessary fields like devicePath
+    devicePath, ok := record.CryptoProvider.Config["devicePath"].(string)
+    if !ok {
+        return nil, fmt.Errorf("device path not found in record")
+    }
+
+    // Initialize the LedgerCryptoProvider with the device path
+    return &LedgerCryptoProvider{DevicePath: devicePath}, nil
+}
+
+// crypto/v2/examples/registerProvider.go
+
+
+import (
+    "crypto/v2/providers"
+    "log"
+)
+
+func main() {
+    // Create an instance of the factory
+    ledgerFactory := &crypto.LedgerCryptoProviderFactory{}
+
+    // Register the factory
+    crypto.RegisterCryptoProviderFactory("LedgerCryptoProvider", ledgerFactory)
```

Comment:
This first parameter here is a string, but the function definition above is of type `CryptoProvider`

Reply:
hmm, I'll give it a look. Thanks!

Reply:
ok, fixed those!